### DISCUSSION TITLE
EVENT RECEIVED TIME DOES NOT MATCH WITH THE LOCAL TIME OF THE SERVER COMPONENT

### DISCUSSION THREAD

**Discussion by dibyendumca** [FreeOpcUa/opcua-asyncio#1094] [Q&A]
Hello,
Can anybody please help?

MY SERVER COMPONENT HAVE THE FOLLOWING CODE SNIPPETS...

 ```
    rootNodeId = server.get_root_node()
	# ...
    # TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT
    # Server Date Time Nodes
    # TTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTTT
    serverLocalTimeNode = await server.nodes.objects.get_child( ["0:Server", "0:LocalTime"])
    serverCurrentTimeNode = await server.nodes.objects.get_child( ["0:Server", "0:ServerStatus", "0:CurrentTime"])
    baseEventLocalTimeNode = await rootNodeId.get_child(["0:Types", "0:ObjectTypes", "0:BaseObjectType", "0:BaseEventType", 
    "0:LocalTime"])
    baseEventReceiveTimeNode = await rootNodeId.get_child(["0:Types", "0:ObjectTypes", "0:BaseObjectType", "0:BaseEventType", 
    "0:ReceiveTime"])
    baseEventTimeNode = await rootNodeId.get_child(["0:Types", "0:ObjectTypes", "0:BaseObjectType", "0:BaseEventType", "0:Time"])
    
	# ...
     await serverLocalTimeNode.set_writable(writable=True)
     await serverCurrentTimeNode.set_writable(writable=True)
     await baseEventLocalTimeNode.set_writable(writable=True)
     await baseEventReceiveTimeNode.set_writable(writable=True)
     await baseEventTimeNode.set_writable(writable=True)
	
	# ...
    # The custom event object automatically will have members from its parent (BaseEventType)
    etype = await server.create_custom_event_type(spaceNodeId, 'MyFirstEvent', ua.ObjectIds.BaseEventType,
                                                  [('MyNumericProperty', ua.VariantType.Int64),
                                                   ('MyStringProperty', ua.VariantType.String)])

    # create second event
    etype2 = await server.create_custom_event_type(spaceNodeId, 'MySecondEvent', ua.ObjectIds.BaseEventType,
                                                   [('MyFloatProperty', ua.VariantType.Float),
                                                    ('MyStringProperty', ua.VariantType.String)])

    # get an event generator for the Param node which generates custom events
    myevgen = await server.get_event_generator(etype, ParamEvents)
    myevgen.event.MyNumericProperty = ua.Variant(-456)
    myevgen.event.MyStringProperty = ua.Variant("First Event")

    # get another event generator for the Param node which generates different custom events
    myevgen2 = await server.get_event_generator(etype2, ParamEvents)
    myevgen2.event.MyFloatProperty = ua.Variant(1.337)
    myevgen2.event.MyStringProperty = ua.Variant("Second Event")

    # get an event generator for the server node which generates BaseEventType
    serverevgen = await server.get_event_generator()
 
   # . . .
   while True:
	TIME = datetime.datetime.now()
	await serverLocalTimeNode.set_value(TIME)
        await serverCurrentTimeNode.set_value(TIME)
	await baseEventLocalTimeNode.set_value(TIME)
	await baseEventReceiveTimeNode.set_value(TIME)
	await baseEventTimeNode.set_value(TIME)
```

AFTER RUNNING THE SERVER COMPONENT, I NOTICED THE TIME BEING UPDATED IN THOSE NODE SPECIFIED.
![image](https://user-images.githubusercontent.com/61493364/196914498-15f3e1d4-3385-4e35-bab3-545837e687c5.png)
![image](https://user-images.githubusercontent.com/61493364/196914558-43456da3-60b5-4467-8684-5ea6e12e2a8f.png)
![image](https://user-images.githubusercontent.com/61493364/196914612-280d07f5-289c-431d-8d2c-5ddd05c2fc3c.png)

I HAVE THE FOLLOWING CODE SNIPPET IN MY CLIENT COMPONET...

```
def event_notification(self, event):
        mevent = Event.get_event_props_as_fields_dict(event)
        print("Event: ", event)

        print("#####################################################")
        print("Source Name       : ", mevent.get("SourceName").Value)
        print("NumericNodeId     : ", mevent.get("EventType").Value)
        print("Event NodeId      : ", ua_utils.val_to_string(
            mevent.get("EventType").Value))

        tagValue = str("''")
        if mevent.keys().__contains__("MyNumericProperty"):
            print("Numeric Property  : ", mevent.get("MyNumericProperty").Value)
            tagValue = "'" + str(mevent.get("MyNumericProperty").Value) + "'"

        if mevent.keys().__contains__("MyFloatProperty"):
            print("Float Property    : ", mevent.get("MyFloatProperty").Value)
            tagValue = "'" + str(mevent.get("MyFloatProperty").Value) + "'"

        if mevent.keys().__contains__("MyStringProperty"):
            print("String Property   : ", mevent.get("MyStringProperty").Value)
            tagValue = "'" + str(mevent.get("MyStringProperty").Value) + "'"

        # receivedTime = mevent.get("Time").Value.strftime(
        #     "%d-%m-%Y %H:%M:%S.%f")

        receivedTime = mevent.get("ReceiveTime").Value.strftime(
            "%d-%m-%Y %H:%M:%S.%f")

        print("Received Time     : ", receivedTime)
        print("#####################################################")
```
**NOW MY QUESTION IS WHY STILL I GET THE UTC TIME, IN THE EVENT NOTIFICATION FUNCTION AS SHOWN BELOW.**

Event: Event(["EventId:b'1b177f41595e4383bf021882771b456a'", 'EventType:NodeId(Identifier=2041, NamespaceIndex=0, NodeIdType=<NodeIdType.FourByte: 1>)', 'SourceNode:NodeId(Identifier=2253, NamespaceIndex=0, NodeIdType=<NodeIdType.FourByte: 1>)', 'SourceName:Server', 'Time:2022-10-20 08:57:36.532349', 'ReceiveTime:2022-10-20 08:57:36.532349', 'LocalTime:TimeZoneDataType(Offset=330, DaylightSavingInOffset=True)', "Message:LocalizedText(Locale=None, Text='Server Event Message 763')", 'Severity:1'])
#####################################################
Source Name : Server
NumericNodeId : NodeId(Identifier=2041, NamespaceIndex=0, NodeIdType=<NodeIdType.FourByte: 1>)
Event NodeId : i=2041
Received Time : 20-10-2022 08:57:36.532349
#####################################################
Event: Event(['MyNumericProperty:-456', 'MyStringProperty:First Event', "EventId:b'765737d8127444a6a83d7642675d163d'", 'EventType:NodeId(Identifier=2, NamespaceIndex=2, NodeIdType=<NodeIdType.FourByte: 1>)', 'SourceNode:NodeId(Identifier=1, NamespaceIndex=2, NodeIdType=<NodeIdType.FourByte: 1>)', 'SourceName:ParamEvents', 'Time:2022-10-20 08:57:37.048961', 'ReceiveTime:2022-10-20 08:57:37.048961', 'LocalTime:TimeZoneDataType(Offset=330, DaylightSavingInOffset=True)', "Message:LocalizedText(Locale=None, Text='This is MyFirstEvent 866')", 'Severity:1'])
#####################################################
Source Name : ParamEvents
NumericNodeId : NodeId(Identifier=2, NamespaceIndex=2, NodeIdType=<NodeIdType.FourByte: 1>)
Event NodeId : ns=2;i=2
Numeric Property : -456
String Property : First Event
Received Time : 20-10-2022 08:57:37.048961
#####################################################
![image](https://user-images.githubusercontent.com/61493364/196915062-13cc3aa0-4aac-4108-8499-fd56f78e3139.png)

**Comment by schroeder-** [comment]
I don't understand what you are doing, but this is not how events are used. The Fields you are setting are only for defining the EventType. Also OPC UA is always in UTC your client has to translate the DateTime values to the clients local time.

**Reply by dibyendumca** [reply]
OK, I see, thanks...

**Reply by dibyendumca** [reply]
But, I want to know, that if I set the Server Component Time to my Local Time, then can I access that time of the Server from my Client component? 
Can the SERVER publish the Current Local Time (if set in the SERVER), and can it be accessed from the CLIENT?
If yes, then HOW ??

**Reply by schroeder-** [reply]
The server updates the servertime every second with the current utc time. You can't override it.

**Reply by dibyendumca** [reply]
That means SERVER TIME cannot be set to Local Current Time

**Reply by schroeder-** [reply]
evgen.trigger(time_attr=TIME)

**Reply by dibyendumca** [reply]
evgen.trigger(time_attr=TIME)

What it does?

**Reply by schroeder-** [reply]
Send the event to the clients? And time_attr is the value for the TIME field.

```Python
myevgen.trigger(time_attr=TIME)
myevgen2.trigger(time_attr=TIME)
serverevgen.trigger(time_attr=TIME)
```

**Reply by dibyendumca** [reply]
How is the Code to retrieve it in the CLIENT Side?

**Reply by schroeder-** [reply]
In event_notification? 
```
def event_notification(self, event):
    mevent = Event.get_event_props_as_fields_dict(event)
    receivedTime = mevent.get("Time").Value.strftime("%d-%m-%Y %H:%M:%S.%f")
```

**Reply by AndreasHeine** [reply]
https://github.com/FreeOpcUa/opcua-asyncio/blob/master/examples/client-events.py

**Reply by dibyendumca** [reply]
OK..... thanks
